### ISSUE TITLE
[MPS] Compiled training is 4x slower than eager because layout optimization forces channels_last

### ISSUE THREAD

**Issue by eyupcanakman** [pytorch/pytorch#192551]
### 🐛 Describe the bug

Training a small conv net on MPS with torch.compile runs 4.5x slower than eager on the current nightly, and 9x slower on the 2.13.0 release wheel.
There are zero graph breaks.
The cause is inductor's layout optimization, and setting `TORCHINDUCTOR_LAYOUT_OPTIMIZATION=0` turns compile into a win at every shape I tried.
I tracked it down with Claude's help and reviewed everything below.
The numbers are from my machine.

Repro, batch 256 at 32x32, bf16 autocast, Apple M5 Pro, macOS 27.

```python
import time
import torch
import torch.nn as nn


class Block(nn.Module):
    def __init__(self, c_in, c_out, stride=1):
        super().__init__()
        self.conv1 = nn.Conv2d(c_in, c_out, 3, stride, 1, bias=False)
        self.bn1 = nn.BatchNorm2d(c_out)
        self.conv2 = nn.Conv2d(c_out, c_out, 3, 1, 1, bias=False)
        self.bn2 = nn.BatchNorm2d(c_out)
        self.short = (
            nn.Sequential(nn.Conv2d(c_in, c_out, 1, stride, bias=False), nn.BatchNorm2d(c_out))
            if stride != 1 or c_in != c_out
            else nn.Identity()
        )

    def forward(self, x):
        out = torch.relu(self.bn1(self.conv1(x)))
        out = self.bn2(self.conv2(out))
        return torch.relu(out + self.short(x))


class Net(nn.Module):
    def __init__(self, w=24):
        super().__init__()
        self.stem = nn.Sequential(nn.Conv2d(3, w, 3, 1, 1, bias=False), nn.BatchNorm2d(w), nn.ReLU())
        self.stage1 = Block(w, w)
        self.stage2 = nn.Sequential(Block(w, 2 * w, 2), Block(2 * w, 2 * w))
        self.stage3 = nn.Sequential(Block(2 * w, 4 * w, 2), Block(4 * w, 4 * w))
        self.head = nn.Linear(4 * w, 10)

    def forward(self, x):
        x = self.stage3(self.stage2(self.stage1(self.stem(x))))
        return self.head(x.mean(dim=(2, 3)))


dev = torch.device("mps")


def bench(compiled, iters=15):
    torch.manual_seed(0)
    m = Net().to(dev)
    if compiled:
        m.compile()
    opt = torch.optim.SGD(m.parameters(), lr=0.1, momentum=0.9)
    lf = nn.CrossEntropyLoss()
    x = torch.randn(256, 3, 32, 32, device=dev)
    y = torch.randint(0, 10, (256,), device=dev)

    def step():
        opt.zero_grad(set_to_none=True)
        with torch.autocast("mps", dtype=torch.bfloat16):
            loss = lf(m(x), y)
        loss.backward()
        opt.step()

    for i in range(iters + 5):
        if i == 5:
            torch.mps.synchronize()
            t = time.time()
        step()
    torch.mps.synchronize()
    return (time.time() - t) / iters * 1000


e = bench(False)
c = bench(True)
print(f"torch {torch.__version__}")
print(f"eager:    {e:7.2f} ms/step")
print(f"compiled: {c:7.2f} ms/step   ({c / e:.2f}x eager)")
```

Output.

```
torch 2.14.0.dev20260807
eager:      25.59 ms/step
compiled:  114.48 ms/step   (4.47x eager)
```

Same model swept across input shapes, eager -> compiled ms/step, with layout optimization on (default) and off (`TORCHINDUCTOR_LAYOUT_OPTIMIZATION=0`), each config in a fresh process and cache.

```
                 layout opt on (default)      TORCHINDUCTOR_LAYOUT_OPTIMIZATION=0
8x8   b256     7.20 ->   9.87   1.37x           7.18 ->  5.58   0.78x
16x16 b256     8.81 ->  31.65   3.59x           7.96 ->  6.75   0.85x
24x24 b128     8.46 ->  35.19   4.16x           8.42 ->  7.11   0.84x
32x32 b128    13.92 ->  60.50   4.35x          13.75 -> 11.95   0.87x
32x32 b256    27.36 -> 119.08   4.35x          27.38 -> 19.78   0.72x
```

With the flag, compiled is faster than eager at every shape.
The regression grows with spatial size, which matches the per-op numbers below.

AI-assisted analysis (Claude), reviewed by me:

`decide_layout_opt` in `torch/_inductor/graph.py` enables channels_last for graphs with convolutions, and `aten.convolution_backward` on MPS runs about 10x slower on channels_last inputs than on contiguous ones at common shapes, so the compiled backward pays that penalty 13 times per step.
With `backend="eager"` the step takes 27.0 ms and with `backend="aot_eager"` 26.5 ms, both matching eager, so the loss is inside inductor codegen rather than dynamo or AOTAutograd.
Timing every call in the generated backward code, with a sync around each call, attributes 84% of the backward to the 13 external `convolution_backward` calls at 8.3 ms per call, while no generated Metal kernel exceeds 1.3 ms.
The op on its own shows the layout penalty in bf16.

```
convolution_backward (256,24,32,32) w=(24,24,3,3)  contiguous 1.84 ms   channels_last 23.52 ms   12.8x
convolution_backward (256,48,16,16) w=(48,48,3,3)  contiguous 0.53 ms   channels_last  5.74 ms   10.9x
convolution_backward (256,96,8,8)   w=(96,96,5,5)  contiguous 1.92 ms   channels_last  3.86 ms    2.0x
```

Forward convolution is unaffected and is even faster on channels_last at these shapes (0.48 ms vs 0.94 ms at 32x32), so inference would not notice.

The heuristics in `decide_layout_opt` come from CUDA benchmarks (the code carries a `TODO - get different values per hardware`) and the function has no MPS branch.
The channels_last penalty inside MPS `convolution_backward` itself may deserve its own look, since it grows with spatial size.

End of the AI-assisted analysis.
I patched `decide_layout_opt` locally to skip layout optimization for MPS training graphs and compiled training went from 4.5x slower to 1.3x faster than eager.
I can send that patch if it is the right direction.

For reference, #150121 tracks compile on MPS but has no conv training performance item, and #189239 is the mirror problem on CUDA and XPU where backward graphs get too little channels_last rather than too much.

### Versions

```
PyTorch version: 2.14.0.dev20260807
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A
OS: macOS 27.0 (arm64)
GCC version: Could not collect
Clang version: 21.0.0 (clang-2100.1.1.101)
CMake version: version 4.4.2
Libc version: N/A
Python version: 3.12.13 (main, Mar  3 2026, 12:39:30) [Clang 21.0.0 (clang-2100.0.123.102)] (64-bit runtime)
Python platform: macOS-27.0-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A
CPU:
Apple M5 Pro
Versions of relevant libraries:
[pip3] numpy==2.5.1
[pip3] torch==2.14.0.dev20260807
[conda] Could not collect
```

cc @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20 @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by Isalia20** [comment]
It reproduces on main. Specifically channels last input is the problem:
```
import time
import torch


WARMUP = 50
ITERATIONS = 200
INPUT_SHAPE = (256, 24, 32, 32)
WEIGHT_SHAPE = (24, 24, 3, 3)


def benchmark(memory_format):
    x = torch.randn(INPUT_SHAPE, device="mps", dtype=torch.bfloat16)
    grad_output = torch.randn(INPUT_SHAPE, device="mps", dtype=torch.bfloat16)
    weight = torch.randn(WEIGHT_SHAPE, device="mps", dtype=torch.bfloat16)
    x = x.contiguous(memory_format=memory_format)
    grad_output = grad_output.contiguous(memory_format=memory_format)
    weight = weight.contiguous(memory_format=memory_format)

    def grad_weight():
        return torch.ops.aten.convolution_backward.default(
            grad_output,
            x,
            weight,
            None,
            (1, 1),
            (1, 1),
            (1, 1),
            False,
            (0, 0),
            1,
            (False, True, False),
        )[1]

    for _ in range(WARMUP):
        result = grad_weight()
    torch.mps.synchronize()

    start = time.perf_counter()
    for _ in range(ITERATIONS):
        result = grad_weight()
    torch.mps.synchronize()
    return (time.perf_counter() - start) * 1000 / ITERATIONS, result.stride()


contiguous_ms, contiguous_stride = benchmark(torch.contiguous_format)
channels_last_ms, channels_last_stride = benchmark(torch.channels_last)

print(f"contiguous: {contiguous_ms:7.2f} ms  stride={contiguous_stride}")
print(f"channels_last: {channels_last_ms:7.2f} ms  stride={channels_last_stride}")
print(f"slowdown: {channels_last_ms / contiguous_ms:7.2f}x")
```

leads to:
```
torch 2.14.0a0+git0754e4e
contiguous:    2.28 ms  stride=(216, 9, 3, 1)
channels_last:   22.14 ms  stride=(216, 1, 72, 24)
slowdown:    9.72x
```

Need to look into this, if it's easily fixable

**Comment by JaynouOliver** [comment]
Opened #193191 with a fix — it skips inductor's layout optimization for MPS training graphs (inference keeps channels_last, which is faster for forward conv). On this issue's benchmark, compiled training goes from 1.5-5x slower than eager to faster than eager at every shape, matching `TORCHINDUCTOR_LAYOUT_OPTIMIZATION=0`; per-shape numbers and the regression test are in the PR. Authored with AI assistance (disclosed in the PR); reviewed and verified locally on Apple Silicon.

@eyupcanakman this follows the same direction as the patch you described — review/feedback welcome. @malfet @Isalia20 could you take a look when you have a moment?

**Comment by ternaus** [comment]
I reviewed these measurements from a real 3D segmentation training route. They are relevant here because `torch.compile` must not implicitly select an MPS layout that makes Conv3d backward slower or invalid.

The benchmark summary below was prepared with AI assistance and reviewed by me:

> Environment: PyTorch 2.13.0, macOS 26.4.1, MPS. Times are per complete training step.
>
> | Layout | Dtype | Eager | `torch.compile` |
> | --- | --- | ---: | ---: |
> | contiguous | FP32 | 41.67 ms | 43.97 ms |
> | contiguous | FP16 | 47.62 ms | 49.82 ms |
> | `channels_last_3d` | FP32 / FP16 | ~42 / ~48 ms | deterministic `aten.convolution_backward` stride assertion |
>
> This extends the report from Conv2d to a representative Conv3d training route. The direction in #193191 is therefore important: Inductor should not automatically move MPS training into a layout that makes backward substantially slower or invalid. On the safe contiguous route, compile is valid but does not improve this workload, and FP16 is slower than FP32.
>
> The `channels_last_3d` assertion appears to be a separate Conv3d/Inductor issue. I will reproduce it against nightly/main before filing a focused follow-up.

#193191 has passed the relevant MPS/Inductor CI; the remaining reported failure is classified as unrelated flakiness. @malfet @Isalia20, could you review the PR and, if this is the accepted direction, mark this issue `actionable`?

**Comment by Isalia20** [comment]
Closing this since it was fixed with https://github.com/pytorch/pytorch/pull/193623
